datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs

```diff
@@ -181,24 +185,30 @@ impl<S: SimplifyInfo> ExprSimplifier<S> {
             expr = expr.rewrite(&mut Canonicalizer::new()).data()?
         }
 
-        // TODO iterate until no changes are made during rewrite
-        // (evaluating constants can enable new simplifications and
-        // simplifications can enable new constant evaluation)
-        // https://github.com/apache/datafusion/issues/1160
-        expr.rewrite(&mut const_evaluator)
-            .data()?
-            .rewrite(&mut simplifier)
-            .data()?
-            .rewrite(&mut guarantee_rewriter)
-            .data()?
-            // run both passes twice to try an minimize simplifications that we missed
-            .rewrite(&mut const_evaluator)
-            .data()?
-            .rewrite(&mut simplifier)
-            .data()?
+        let mut i = 0;
+        loop {
+            let result = expr.rewrite(&mut const_evaluator)?;
+            let mut transformed = result.transformed;
+            expr = result.data;
+
+            let result = expr.rewrite(&mut simplifier)?;
+            transformed |= result.transformed;
+            expr = result.data;
+
+            let result = expr.rewrite(&mut guarantee_rewriter)?;
+            transformed |= result.transformed;
+            expr = result.data;
+
             // shorten inlist should be started after other inlist rules are applied
-            .rewrite(&mut shorten_in_list_simplifier)
-            .data()
+            let result = expr.rewrite(&mut shorten_in_list_simplifier)?;
+            transformed |= result.transformed;
+            expr = result.data;
+
+            i += 1;
+            if !transformed || i >= self.max_simplifier_iterations {
+                return Ok(expr);
```

Comment:
returning `(Expr, usize)` would be nice for testing. Especially if the number of iterations becomes a configurable parameter - which opens the door for new simplification bugs - we might want to test that specifically X number of iterations runs on a given expression.

Reply:
Maybe we could add a function like `simplify_inner` that did and then use that in tests

```rust
    pub fn simplify(&self, mut expr: Expr) -> Result<Expr> {
       self.simplify_inner(expr).0
    }

   /// Returns the simplified expr and the number of iterations required.
   fn simplify_inner(&self, mut expr: Expr) -> Result<(usize, Expr)> {
```